sdk/iot/hub/samples/twin/src/iot_hub_twin_example.c

```diff
@@ -0,0 +1,447 @@
+{
+  az_result result;
+  result = az_json_builder_init(json_builder, AZ_SPAN_FROM_BUFFER(reported_property_payload));
+  result = az_json_builder_append_token(json_builder, az_json_token_object_start());
+  az_json_token reported_property_value_token
+      = az_json_token_number((double)reported_property_value++);
+  result = az_json_builder_append_object(
+      json_builder, reported_property_name, reported_property_value_token);
+  result = az_json_builder_append_token(json_builder, az_json_token_object_end());
+
+  return result;
+}
+
+static int send_reported_property()
+{
+  int rc;
+  printf("Sending reported property\n");
+
+  //Get the topic if we haven't already
+  if (reported_property_topic[0] == '\0')
+  {
+    if ((rc = az_iot_hub_client_twin_patch_get_publish_topic(
+             &client,
+             reported_property_topic_request_id,
+             reported_property_topic,
+             sizeof(reported_property_topic),
+             NULL))
+        != AZ_OK)
+    {
+      printf("Unable to get twin document publish topic, return code %d\n", rc);
+      return rc;
+    }
+  }
+
+  az_json_builder json_builder;
+  if((rc = build_reported_property(&json_builder)) != AZ_OK)
+  {
+    return rc;
+  }
+  az_span json_payload = az_json_builder_span_get(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           reported_property_topic,
+           az_span_size(json_payload),
+           az_span_ptr(json_payload),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish reported property, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+int main()
+{
+  int rc;
+
+  if ((rc = read_configuration_and_init_client()) != AZ_OK)
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  char client_id[128];
+  if ((rc = az_iot_hub_client_get_client_id(&client, client_id, sizeof(client_id), NULL)) != AZ_OK)
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_create(
+           &mqtt_client, mqtt_endpoint, client_id, MQTTCLIENT_PERSISTENCE_NONE, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to create MQTT client, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to set MQTT callbacks, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = connect_device()) != 0)
+  {
+    return rc;
+  }
+
+  if ((rc = subscribe()) != 0)
+  {
+    return rc;
+  }
+
+  printf("Subscribed to topics.\n");
+
+  printf("\nWaiting for activity:\nPress 'g' to get the twin document\nPress 'r' to send a reported "
+         "property\n[Press 'q' to quit]\n");
+
+  int input;
+  while (1)
```

Comment:
We shouldn't have IoT samples interact with the user. It's unusual to have IoT devices requiring user input.

Reply:
Noted. Will update in a later PR.